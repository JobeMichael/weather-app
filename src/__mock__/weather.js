export const weather = {
  dt: 1597989600,
  main: {
    temp: 22.4,
    feels_like: 22.19,
    temp_min: 22.4,
    temp_max: 23.43,
    pressure: 1011,
    sea_level: 1012,
    grnd_level: 1007,
    humidity: 67,
    temp_kf: -1.03,
  },
  weather: [
    {
      id: 803,
      main: "Clouds",
      description: "broken clouds",
      icon: "04d",
    },
  ],
  clouds: {
    all: 70,
  },
  wind: {
    speed: 3.12,
    deg: 161,
  },
  visibility: 10000,
  pop: 0,
  sys: {
    pod: "d",
  },
  dt_txt: "2020-08-21 06:00:00",
};
