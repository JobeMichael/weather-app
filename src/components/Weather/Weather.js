import { useFetch } from "hooks/useFetch";
import moment from "moment";
import React from "react";
import { getIconUrl } from "utils/getIconUrl";
import { getTimeFromUTC } from "utils/getTimeFromUTC";
import { getUrl } from "utils/getUrl";
import * as Styled from "./Weather.styles";
import WeatherDetailCell from "./WeatherDetailCell/WeatherDetailCell";

const Weather = ({ position }) => {
  const weatherApiUrl = getUrl("weather", position);
  const { data, loading, error } = useFetch(weatherApiUrl);

  if (error) {
    return <></>;
  }

  if (loading || !data) {
    return <></>;
  }

  const {
    weather: [{ description, icon }],
    main: { temp, feels_like, temp_min, temp_max, humidity },
    wind: { speed: windSpeed },
    sys: { country, sunrise, sunset },
    name,
  } = data;

  return (
    <Styled.Wrapper>
      <Styled.Header>
        <h2>
          {name}, {country}
        </h2>
        <h4>{moment(new Date()).format("dddd D, MMMM YYYY")}</h4>
      </Styled.Header>
      <Styled.Temperature>
        <div>
          <img src={getIconUrl(icon)} alt="weather-icon"></img>
        </div>
        <div>
          <h3>{Math.floor(temp)}째</h3>
          <p>Feels like {Math.floor(feels_like)}째</p>
          <h4>{description}</h4>
        </div>
      </Styled.Temperature>
      <Styled.WeatherDetailsWrapper>
        <WeatherDetailCell value={`${Math.floor(temp_max)}째`} title="Hight" />
        <WeatherDetailCell value={`${windSpeed}mph`} title="Wind" />
        <WeatherDetailCell value={getTimeFromUTC(sunrise)} title="Sunrise" />
        <WeatherDetailCell value={`${Math.floor(temp_min)}째`} title="Low" />
        <WeatherDetailCell value={`${humidity}%`} title="Humidity" />
        <WeatherDetailCell value={getTimeFromUTC(sunset)} title="Sunset" />
      </Styled.WeatherDetailsWrapper>
    </Styled.Wrapper>
  );
};

export default Weather;
