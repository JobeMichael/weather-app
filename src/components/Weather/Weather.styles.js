import styled from "styled-components";

export const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
`;

export const Header = styled.div`
  flex-basis: 100%;
  h2 {
    color: rgb(255, 255, 255);
    display: block;
    font-weight: 600;
    font-size: 30px;
    text-align: center;
    padding: 5px 0px;
  }
  h4 {
    color: rgb(255, 255, 255);
    display: block;
    font-weight: 400;
    font-size: 15px;
    text-align: center;
    padding: 5px 0px;
  }
`;

export const WeatherDetailsWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  background-color: rgba(255, 255, 255, 0.2);
  align-self: flex-start;
  padding: 10px 0px;
  margin: 20px 0px;
  border-radius: 10px;
  flex-basis: 50%;
`;

export const Temperature = styled.div`
  display: flex;
  align-items: center;
  h3 {
    font-size: 80px;
  }
  h4 {
    font-size: 30px;
  }
`;
