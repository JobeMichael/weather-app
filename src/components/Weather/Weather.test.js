import { cleanup, render } from "@testing-library/react";
import React from "react";
import { weather as mockData } from "__mock__/weather";
import { useFetch } from "../../hooks/useFetch";
import Weather from "./Weather";

jest.mock("../../hooks/useFetch");

afterEach(cleanup);

describe("<Weather/>", () => {
  it("renders Weather correctly", () => {
    useFetch.mockReturnValue({
      loading: false,
      data: mockData,
    });
    const { asFragment } = render(<Weather position={{ lat: 1, lon: 3 }} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should show Loader during async call", () => {
    useFetch.mockReturnValue({
      loading: true,
      data: null,
    });
    const { asFragment } = render(<Weather position={{ lat: 1, lon: 3 }} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should show error message if there is a API error", () => {
    useFetch.mockReturnValue({
      loading: true,
      error: "Error",
    });
    const { asFragment } = render(<Weather position={{ lat: 1, lon: 3 }} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
