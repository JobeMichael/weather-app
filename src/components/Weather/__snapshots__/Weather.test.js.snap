// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Weather/> renders Weather correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm iIkEOg"
  >
    <div
      class="sc-AxirZ jymRuQ"
    >
      <h2>
        , 
      </h2>
      <h4>
        Friday 21, August 2020
      </h4>
    </div>
    <div
      class="sc-AxhCb kurrzz"
    >
      <div>
        <img
          alt="weather-icon"
          src="http://openweathermap.org/img/wn/04d@4x.png"
        />
      </div>
      <div>
        <h3>
          22째
        </h3>
        <p>
          Feels like 22째
        </p>
        <h4>
          broken clouds
        </h4>
      </div>
    </div>
    <div
      class="sc-AxiKw kjKCal"
    >
      <div
        class="sc-AxhUy bDbuoR"
      >
        <h4>
          23째
        </h4>
        <span>
          Hight
        </span>
      </div>
      <div
        class="sc-AxhUy bDbuoR"
      >
        <h4>
          3.12mph
        </h4>
        <span>
          Wind
        </span>
      </div>
      <div
        class="sc-AxhUy bDbuoR"
      >
        <h4>
          Inval
        </h4>
        <span>
          Sunrise
        </span>
      </div>
      <div
        class="sc-AxhUy bDbuoR"
      >
        <h4>
          22째
        </h4>
        <span>
          Low
        </span>
      </div>
      <div
        class="sc-AxhUy bDbuoR"
      >
        <h4>
          67%
        </h4>
        <span>
          Humidity
        </span>
      </div>
      <div
        class="sc-AxhUy bDbuoR"
      >
        <h4>
          Inval
        </h4>
        <span>
          Sunset
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Weather/> should show Loader during async call 1`] = `
<DocumentFragment>
  <h3>
    Loading
  </h3>
</DocumentFragment>
`;

exports[`<Weather/> should show error message if there is a API error 1`] = `
<DocumentFragment>
  <h1>
    Error...
  </h1>
</DocumentFragment>
`;
